compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -2228,24 +2228,32 @@ private void validateFieldsAndSetReadOnlyType(List<BLangNode> typeDefNodes, Symb
                 continue;
             }
 
-            boolean allReadOnlyOrFinalFields = true;
+            boolean allImmutableFields = true;
 
             Collection<BField> fields = structureType.fields.values();
 
             if (fields.isEmpty()) {
                 continue;
             }
 
-            int flagToCheck = nodeKind == NodeKind.RECORD_TYPE ? Flags.READONLY : Flags.FINAL;
-
-            for (BField field : fields) {
-                if (!Symbols.isFlagOn(field.symbol.flags, flagToCheck)) {
-                    allReadOnlyOrFinalFields = false;
-                    break;
+            if (nodeKind == NodeKind.RECORD_TYPE) {
+                for (BField field : fields) {
+                    if (!Symbols.isFlagOn(field.symbol.flags, Flags.READONLY)) {
+                        allImmutableFields = false;
+                        break;
+                    }
+                }
+            } else {
+                for (BField field : fields) {
+                    if (!Symbols.isFlagOn(field.symbol.flags, Flags.FINAL) ||
+                            !Symbols.isFlagOn(structureType.flags, Flags.READONLY)) {
```

Comment:
:pick: Extract this out of for loop? `Symbols.isFlagOn(structureType.flags, Flags.READONLY)`

Reply:
We don't actually need this check here now since object type descriptor fields cannot be marked as `final`. Removed in e2badab.